presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterIntoWindow.java

```diff
@@ -0,0 +1,184 @@
+    private final Pattern<FilterNode> pattern;
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownFilterIntoWindow(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+        this.childCapture = newCapture();
+        this.pattern = filter()
+                .with(source().matching(window()
+                        .matching(window -> window.getWindowFunctions().size() == 1)
+                        .matching(window -> getOnlyElement(window.getWindowFunctions().values()).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId))
+                        .capturedAs(childCapture)));
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return pattern;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeTopNRowNumber(session);
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+
+        WindowNode windowNode = captures.get(childCapture);
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+        OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+        if (upperBound.isPresent()) {
+            if (upperBound.getAsInt() <= 0) {
+                return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+            }
+            if (windowNode.getOrderingScheme().isPresent()) {
+                TopNRowNumberNode newSource = new TopNRowNumberNode(idAllocator.getNextId(),
+                        windowNode.getSource(),
+                        windowNode.getSpecification(),
+                        getOnlyElement(windowNode.getWindowFunctions().keySet()),
+                        upperBound.getAsInt(),
+                        false,
+                        Optional.empty());
+
+                if (!allRowNumberValuesInDomain(tupleDomain, rowNumberSymbol, upperBound.getAsInt())) {
+                    return Result.ofPlanNode(new FilterNode(node.getId(), newSource, node.getPredicate()));
+                }
+
+                // Remove the row number domain because it is absorbed into the node
+                DomainTranslator.ExtractionResult extractionResult = fromPredicate(metadata, session, node.getPredicate(), types);
```

Comment:
Please reuse the `ExtractionResult` (like you did in `PushdownFilterIntoRowNumber` rule)

Reply:
Do you mean we can move `ExtractionResult` at the top of the `apply` method?